pkg/ruler/namespace_protection.go

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ruler
+
+import (
+	"context"
+	"errors"
+	"strings"
+)
+
+const (
+	OverrideProtectionHeader  = "X-Mimir-Ruler-Override-Namespace-Protection"
+	ProtectedNamespacesHeader = "X-Mimir-Ruler-Protected-Namespaces"
+)
+
+var (
+	ErrNoProtectionOverrideHeader = errors.New("no protection override header")
+	ErrNamespaceTargetNotMatch    = errors.New("namespace does not match target")
+)
+
+// IsNamespaceProtected returns true if the namespace is protected for the given user.
+func (r *Ruler) IsNamespaceProtected(userID string, namespace string) bool {
```

Comment:
Do you need all these functions to be exposed?

Reply:
Discussed this offline, the answer is yes.